From 3af2c38ed41122d10b576608d567b2b94b71ddcf Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Fri, 16 Dec 2022 14:52:42 +0530
Subject: [PATCH 2/4] cfg80211/mac80211: update beacon counters per link basis

Currently, function to update beacon counter uses deflink to fetch
the beacon and then update the counter. However, with MLO, there is
a need to update the counter for the beacon in a particular link.

Add support to use link_id in order to fetch the beacon from a particular
sdata during beacon update counter.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c |  2 +-
 drivers/net/wireless/ath/ath12k/wmi.c |  3 ++-
 6 files changed, 20 insertions(+), 9 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1622,7 +1622,7 @@ void ath12k_mac_bcn_tx_event(struct ath1
 	arvif->bcca_zero_sent = false;
 
 	if (link_conf->color_change_active)
-		ieee80211_beacon_update_cntdwn(vif);
+		ieee80211_beacon_update_cntdwn(vif, 0);
 	ieee80211_queue_work(ar->ah->hw, &arvif->update_bcn_template_work);
 }
 
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -8054,7 +8054,8 @@ ath12k_wmi_process_csa_switch_count_even
 			if (!ev->current_switch_count)
 				ieee80211_csa_finish(arvif->ahvif->vif, arvif->link_id);
 			else if (ev->current_switch_count > 1)
-				ieee80211_beacon_update_cntdwn(arvif->ahvif->vif);
+				ieee80211_beacon_update_cntdwn(arvif->ahvif->vif,
+							       arvif->link_id);
 		}
 	}
 	rcu_read_unlock();
